enhancements/etcd/scaling-etcd-with-raft-learners.md

```diff
@@ -0,0 +1,343 @@
+
+### Cluster Member Controller
+
+The cluster member controller manages scaling up etcd during member replacement and bootstrap. This proposal makes
+changes to the way this controller manages scale up.
+
+`ensureEtcdLearnerPromotion`: This new method will provide the logic necessary to ask the cluster for the list of
+members and attempt to promote any learner members which have a log in sync with the leader.
+
+#### Reads
+
+1. openshift-etcd pods/quorum-guard: Today the controller loops not `Ready` etcd pods and will attempt to scale
+up if the etcd is not already part of the cluster membership. While this is effective during bootstrap it is not
+efficient as scaling no longer needs to be serial. Also by ensuring the quorum guard pod exists on the node before
+scale up we can give more power to the PDB in scaling.
+
+2. `MemberListResponse`: The Cluster API `MemberListRequest` RPC provides `[]Member` for the etcd cluster including the
+method `IsLearner`.
+
+- Consumers: scale up and promotion of learner members
+
+- ref: https://pkg.go.dev/github.com/coreos/etcd/etcdserver/etcdserverpb#MemberListResponse
+
+3. `MemberPromoteResponse`: The Cluster API `MemberPromoteRequest` asks the leader of the term if the learner's log
+is in sync with its own. If no return error.
+
+- Consumers `ensureEtcdLearnerPromotion`
+
+- ref: https://pkg.go.dev/go.etcd.io/etcd/api/v3/etcdserverpb#MemberPromoteResponse
+
+### Risks and Mitigations
+
+The largest risks to scaling is quorum loss, data loss, and split brain scenario.
+
+- **Quorum Loss**: Because new members are added as non voting members and can not be promoted to voting members unless
+the etcd process starts and the log has been completely replicated and in sync with leader.
+
+
+- **Split brain**: The cluster gates starting of the etcd process on a verification process based on the cluster
+member id. This ensures that each revision has an explicit expected membership. Because quorum guard replicas are
+managed by the operator the cluster topology will remain in a safe configuration (odd number of members).
+
+
+- **Data loss**: A concern with rolling scaling of etcd with large data files is possible data loss. The MTTR of log
+replication is directly tied to the size of the state. If members of the cluster were replaced too quickly in
+conjunction of it could be possible although very unlikely that no member has a complete log. Raft learners ensures
+this is not possible by waiting for the log to be replicated from the leader before promotion to voting member can
+take place.
+
+## Design Details
+
+### Open Questions [optional]
+
+This is where to call out areas of the design that require closure before deciding to implement the
+design.  For instance, > 1. This requires exposing previously private resources which contain
+sensitive information.  Can we do this?
+
+### Test Plan
+
+- e2e with
+
+    1. add dangling etcd learner member which has not been started and not promoted.
```

Comment:
Curious how will we prevent a learner from being promoted? Otherwise sounds great!

Reply:
One idea is to use IPTables to block peer traffic` tcp` port `2380`.